framework/Optimizers/SimulatedAnnealing.py

```diff
@@ -0,0 +1,718 @@
+        self.raiseAMessage('Optimizer "{}" trajectory {} was unable to continue due to functional or boundary constraints.'
+                          .format(self.name, traj))
+        self._closeTrajectory(traj, 'converge', 'no constraint resolution', newPoint[self._objectiveVar])
+        return
+      self._submitRun(suggested, traj, self.getIteration(traj))
+
+  # * * * * * * * * * * * * * * * *
+  # Convergence Checks
+  # Note these names need to be formatted according to checkConvergence check!
+  convFormat = ' ... {name:^12s}: {conv:5s}, {got:1.2e} / {req:1.2e}'
+
+  # NOTE checkConvSamePoint has a different call than the others
+  # should this become an informational dict that can be passed to any of them?
+
+  def checkConvergence(self, traj, new, old):
+    """
+      Check for trajectory convergence
+      @ In, traj, int, trajectory to consider
+      @ Out, None? FIXME
+    """
+    convs = {}
+    for conv in self._convergenceCriteria:
+      # special treatment for same point check
+      if conv == 'samePoint':
+        convs[conv] = self._checkConvSamePoint(new, old)
+        continue
+      # fix capitalization for RAVEN standards
+      fName = conv[:1].upper() + conv[1:]
+      # get function from lookup
+      f = getattr(self, '_checkConv{}'.format(fName))
+      # check convergence function
+      okay = f(traj)
+      # store and update
+      convs[conv] = okay
+    return any(convs.values()), convs
+
+  def _checkConvSamePoint(self, new, old):
+    """
+      Checks for a repeated same point
+      @ In, new, dict, new opt point
+      @ In, old, dict, old opt point
+      @ Out, converged, bool, convergence state
+    """
+    # TODO diff within tolerance? Exactly equivalent seems good for now
+    same = list(abs(new[var] - old[var])==self._convergenceCriteria['samePoint'] for var in self.toBeSampled)
+    converged = all(same)
+    self.raiseADebug(self.convFormat.format(name='same point',
+                                            conv=str(converged),
+                                            got=sum(same),
+                                            req=len(same)))
+    return converged
+  def _checkConvObjective(self, traj):
+    """
+      Checks the change in objective for convergence
+      @ In, traj, int, trajectory identifier
+      @ Out, converged, bool, convergence state
+    """
+    if len(self._optPointHistory[traj]) < 2:
+      return False
+    o1, _ = self._optPointHistory[traj][-1]
+    o2, _ = self._optPointHistory[traj][-2]
+    delta = o2[self._objectiveVar]-o1[self._objectiveVar]
+    converged = abs(delta) < self._convergenceCriteria['objective']
+    self.raiseADebug(self.convFormat.format(name='objective',
+                                            conv=str(converged),
+                                            got=delta,
+                                            req=self._convergenceCriteria['objective']))
+    return converged
+
+  def _checkConvTemperature(self, traj):
```

Comment:
where do you use it ? i can not find it

Reply:
fName = conv[:1].upper() + conv[1:]
f = getattr(self, '_checkConv{}'.format(fName))
okay = f(traj)

Reply:
I do not understand the answer... the method ```_checkConvTemperature``` is not called by any other method in this module... where do you use it?

Reply:
It is called in the lines I pasted here. when `fName` is 'objective' then `okay = f(traj)` is just _checkConvObjective, and then fName is 'Temperature' then 
`okay = f(traj)` will call _checkConvTemperature() and so on.